### PR TITLE
chore: smaller view models [AR-3252]

### PR DESCRIPTION
**Pull request by Garzas**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-3252" title="AR-3252" target="_blank">TaskAR-3252</a>  Implement ViewModels for Smaller Composables
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Context
As our Android Reloaded Wire app continues to grow in complexity, we have encountered some challenges with maintaining large view models. This has led us to explore strategies for breaking down our view models into smaller, more manageable components.

### Decision 
We have decided to adopt the approach outlined in the article "Divide big view models to smaller" by Q42 Engineering, and implement it using the Resaca library for breaking down our view models into smaller components.

### Alternatives
We considered several alternative approaches, including using interfaces with functions for each particular composable, which were then passed in the composable tree. However, we ultimately decided that the approach outlined in the Q42 article and implemented with Resaca was the most suitable for our needs.

### Consequences
By dividing our large view models into smaller components, we expect to see improvements in several areas, including code organization, testability, maintainability, and reusability of the composables. Smaller view models will allow us to create more focused and targeted components, which can be reused across multiple screens and layouts, reducing the amount of duplicate code in our codebase. 

### Status
We have already begun implementing this approach in our Android Reloaded Wire app using the Resaca library for breaking down our view models into smaller components. We will continue to monitor its effectiveness and make adjustments as needed.

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_ 

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/2/) **failed**.

File: app/src/main/kotlin/com/wire/android/model/LoadableState.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * Wrapper for view model states with additional [isLoading] variable for UI purposes
+ */
+data class LoadableState<T>(
+    val state: T,
+    val isLoading: Boolean = false
+)
+
+fun <T> LoadableState<T>.startLoading(): LoadableState<T> = this.copy(isLoading = true)
+fun <T> LoadableState<T>.finishLoading(): LoadableState<T> = this.copy(isLoading = false)
+fun <T> LoadableState<T>.updateState(newState: T): LoadableState<T> = this.copy(state = newState, isLoading = false)
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/model/LoadableState.kt]: With `state` being mandatory, what do we pass while it's still loading?

For example, representing the state of a user profile.
```kt
LoadableState(
    state = ????,
    isLoading = true
)
```

Maybe a sealed class?

```kotlin
sealed interface LoadableState<T> {

    data object Loading<T>: LoadableState<T> // Maybe we could have a percentage in the future just for loading?

    data class Completed(val data: T): LoadableState<T>

    // Maybe error state?
    data class Failure(val cause: CoreFailure): LoadableState<T>
}
```

Pretty much the current `TriState` we already have.

Why use this and not the `TriState`?

File: app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButton.kt [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+@Composable
+fun ConnectionActionButton(
+    userId: UserId,
+    userName: String,
+    connectionStatus: ConnectionState
+) {
+    val viewModel: ConnectionActionButtonBaseViewModel = if (LocalInspectionMode.current) {
+        ConnectionActionButtonPreviewModel(LoadableState(connectionStatus))
+    } else {
+        hiltViewModelScoped<ConnectionActionButtonViewModel>(
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButton.kt]: Here the Composable needs to manage:

- BaseViewModel
- ViewModel
- PreviewModel

It's a lot of boilerplate code.

I'd prefer if we could generate this automatically some how. I'll give it some thought, as this is an improvement that can be done later.

On a similar note, what if we keep the naming like we already do with other instances:

```kotlin
interface FooViewModel
class FooViewModelImpl: FooViewModel
class FooViewModelPreview: FooViewModel
```

So the default `ConnectionActionButtonViewModel` is actually the interface that the composables can use, and the implementation is explicitly marked as such.

**Review by vitorhugods — COMMENTED** [review]
In general, I like it :)

Just a couple of topics to discuss.

File: app/src/main/kotlin/com/wire/android/model/LoadableState.kt (hunk 1 above)
Comment (Garzas) [reply app/src/main/kotlin/com/wire/android/model/LoadableState.kt]: We don't need to propagate Error state to UI which may complicate it

File: app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButton.kt (hunk 2 above)
Comment (Garzas) [reply app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButton.kt]: With naming I agree :)

File: app/src/main/kotlin/com/wire/android/ui/userprofile/other/OtherUserProfileScreen.kt [hunk 3]
```diff
@@ -106,18 +108,19 @@ fun OtherUserProfileScreen(viewModel: OtherUserProfileScreenViewModel = hiltView
...
-        requestInProgress = viewModel.requestInProgress,
-        sheetState = sheetState,
-        openBottomSheet = openBottomSheet,
-        closeBottomSheet = closeBottomSheet,
-        snackbarHostState = snackbarHostState,
-        eventsHandler = viewModel,
-        footerEventsHandler = viewModel,
-        bottomSheetEventsHandler = viewModel
-    )
+    CompositionLocalProvider(LocalSnackbarHostState provides snackbarHostState) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/userprofile/other/OtherUserProfileScreen.kt]: I think it woukld make sense to provide it from the very first screen which would be WireActivity

**Review by trOnk12 — APPROVED** [review]
What is the action plan with smaller view models as for now ? Will they not be handeling snackbar /modalbottom sheet and stuff ?

File: app/src/main/kotlin/com/wire/android/model/LoadableState.kt (hunk 1 above)
Comment (vitorhugods) [reply app/src/main/kotlin/com/wire/android/model/LoadableState.kt]: Then, what about a `DataState`? Just like the `TriState` but without the `Failure` part.

This way we don't need to pass a made-up state (or nullable state) when it's still loading.

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.2.0(6587424)(6587424)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.2.0(6587424)(6587424)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/model/LoadableState.kt (hunk 1 above)
Comment (Garzas) [reply app/src/main/kotlin/com/wire/android/model/LoadableState.kt]: We constructed those UI to show previous state when request is loading

File: app/src/main/kotlin/com/wire/android/ui/userprofile/other/OtherUserProfileScreen.kt (hunk 3 above)
Comment (Garzas) [reply app/src/main/kotlin/com/wire/android/ui/userprofile/other/OtherUserProfileScreen.kt]: I agree

**Comment by Garzas** [comment]
> What is the action plan with smaller view models as for now ? Will they not be handeling snackbar /modalbottom sheet and stuff ?

It will but it the logic doesn't change for dialogs or bottom sheets, it's strictly connected with each composable so it doesn't matter what is the parent of this composable (dialog, bottom sheet or screen)

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.2.0(6587965)(6587965)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.2.0(6587965)(6587965)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButtonViewModelImpl.kt
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButtonViewModelImpl.kt:1]: Rename this file to `ConnectionActionButtonViewModel`? :D

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/7/) **failed**.

**Review by vitorhugods — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/8/) **failed**.

**Comment by AndroidBob** [comment]
Build [9](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/9/) **failed**.

**Comment by AndroidBob** [comment]
Build [10](https://10.10.124.134/job/wire-android-reloaded/job/PR-1810/10/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.2.0(6594881)(6594881)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.2.0(6594881)(6594881)-staging-compat.apk)

**Pull request merged**